### PR TITLE
Add support for SRV record for ClusterIP services 

### LINKED ISSUE #526: Support for SRV records
**What would you like to be added**:
Support SRV record in Lighthouse for exported services

**Why is this needed**:
SRV records help in discovering services by describing the protocol/s and address of headless services

More details in: https://github.com/submariner-io/enhancements/blob/devel/lighthouse/srv-records.md

### PR DESCRIPTION
**Pull request by aswinsuryan**
Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

### CODE REVIEW

File: pkg/service/controller.go [hunk 1]
```diff
@@ -84,24 +87,41 @@ func (c *Controller) Stop() {
...
+		}
+	}
+
+	record := &serviceimport.DNSRecord{
+		IP:   svc.Spec.ClusterIP,
+		Port: mcsServicePort,
+	}
 	if svc.Spec.Type == v1.ServiceTypeClusterIP && svc.Spec.ClusterIP != "" {
-		return svc.Spec.ClusterIP, true
+		return record, true
```
Comment (tpantelis) [inline pkg/service/controller.go]: You can move this check above L105:

```
if svc.Spec.Type != v1.ServiceTypeClusterIP || svc.Spec.ClusterIP == "" {
    return nil, false
}
```

File: pkg/serviceimport/map.go [hunk 2]
```diff
@@ -23,24 +23,28 @@ import (
 	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 )
 
+type DNSRecord struct {
+	IP   string
+	Port []mcsv1a1.ServicePort
+}
```
Comment (tpantelis) [inline pkg/serviceimport/map.go:31]: New line.

```suggestion
}

```

File: pkg/serviceimport/map.go [hunk 3]
```diff
@@ -61,16 +65,16 @@ func (m *Map) selectIP(queue []clusterInfo, counter *uint64, name, namespace str
...
 
-	return ""
+	return nil
 }
 
 func (m *Map) GetIP(namespace, name, cluster, localCluster string, checkCluster func(string) bool,
-	checkEndpoint func(string, string, string) bool) (ip string, found, isLocal bool) {
-	clusterIPs, queue, counter, isHeadless := func() (map[string]string, []clusterInfo, *uint64, bool) {
+	checkEndpoint func(string, string, string) bool) (record *DNSRecord, found, isLocal bool) {
+	clusterIPs, queue, counter, isHeadless := func() (map[string]*DNSRecord, []clusterInfo, *uint64, bool) {
```
Comment (tpantelis) [inline pkg/serviceimport/map.go]:
```suggestion
	dnsRecords, queue, counter, isHeadless := func() (map[string]*DNSRecord, []clusterInfo, *uint64, bool) {
```

File: plugin/lighthouse/handler.go [hunk 4]
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
...
+		record := &dns.A{Hdr: dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass(), Ttl: lh.ttl}, A: net.ParseIP(ip).To4()}
+		log.Debugf("rr is %v", record)
+		records = append(records, record)
+	}
+
+	return records
+}
+
+func (lh *Lighthouse) createSRVRecords(record *serviceimport.DNSRecord, state request.Request, pReq recordRequest, zone string) []dns.RR {
+	reqPorts := make([]mcsv1a1.ServicePort, 0)
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]: No need to initialize if not needed.

```suggestion
	var reqPorts []mcsv1a1.ServicePort
```

File: plugin/lighthouse/handler.go [hunk 5]
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
...
+	records := make([]dns.RR, 0)
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]: Move this right before L110:

```suggestion
	records := make([]dns.RR, len(reqPorts))
```

File: plugin/lighthouse/handler.go [hunk 6]
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
 		return lh.nextOrFailure(state.Name(), ctx, w, r, dns.RcodeNameError, "Only services supported")
 	}
 
+	return lh.getDNSRecord(zone, state, ctx, w, r, pReq)
+}
+
+func (lh *Lighthouse) createARecords(ips []string, state request.Request) []dns.RR {
+	records := make([]dns.RR, 0)
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]:
```suggestion
	records := make([]dns.RR, 0, len(ips))
```

File: plugin/lighthouse/handler.go [hunk 7]
```diff
@@ -91,12 +152,18 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
 
 	records := make([]dns.RR, 0)
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]:
```suggestion
	var records []dns.RR
```

File: plugin/lighthouse/handler.go [hunk 8]
```diff
@@ -91,12 +152,18 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
...
 
-	for _, ip := range ips {
-		record := &dns.A{Hdr: dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass(), Ttl: lh.ttl}, A: net.ParseIP(ip).To4()}
-		log.Debugf("rr is %v", record)
-		records = append(records, record)
+	if state.QType() == dns.TypeA {
+		records = lh.createARecords(ips, state)
+	} else if state.QType() == dns.TypeSRV {
+		records = lh.createSRVRecords(record, state, pReq, zone)
+		if records == nil {
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]: Perhaps this should be moved after L165?

```suggestion
		if len(records) {
```

File: plugin/lighthouse/handler.go [hunk 9]
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
...
+	}
+
+	for _, port := range record.Port {
+		portKey := strings.ToLower(port.Name) + strings.ToLower(string(port.Protocol))
+		log.Debugf("The current port Name %q SRV require port Name %q", portKey, portReqKey)
+
+		if portKey == portReqKey {
+			reqPorts = append(reqPorts, port)
+		}
+	}
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]: If no port is requested then we're unnecessarily looping on `record.Port`.  Also what if` pReq.port` is empty but `pReq.protocol` isn't?  How about: 

```suggestion
	if pReq.port == "" {
		reqPorts = record.Port
	} else {
	    log.Debugf("Requested port %q, protocol %q for SRV", pReq.port, pReq.protocol)
	    for _, port := range record.Port {
		name := strings.ToLower(port.Name)
		protocol := strings.ToLower(string(port.Protocol))
		
		log.Debugf("Checking port %q, protocol %q", name, protocol)

		if name == pReq.port && protocol ==  pReq.protocol {
			reqPorts = append(reqPorts, port)
		}
	    }
	}
```

File: plugin/lighthouse/handler.go [hunk 10]
```diff
@@ -61,26 +65,83 @@ func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
...
+		return nil
+	}
+
+	target := pReq.service + "." + pReq.namespace + "." + zone
+
+	if pReq.cluster != "" {
+		target = pReq.cluster + "." + target
+	}
+
+	for _, ports := range reqPorts {
```
Comment (tpantelis) [inline plugin/lighthouse/handler.go]:
```suggestion
	for _, port := range reqPorts {
```

File: pkg/serviceimport/map.go [hunk 11]
```diff
@@ -23,24 +23,28 @@ import (
 	mcsv1a1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 )
 
+type DNSRecord struct {
+	IP   string
+	Port []mcsv1a1.ServicePort
```
Comment (tpantelis) [inline pkg/serviceimport/map.go]:
```suggestion
	Ports []mcsv1a1.ServicePort
```

File: plugin/lighthouse/parse.go [hunk 12]
```diff
@@ -115,3 +107,38 @@ func (r recordRequest) String() string {
...
+	// Because of ambiguity we check the labels left: 1: a cluster. 2: hostname and cluster.
+	// Anything else is a query that is too long to answer and can safely be delegated to return an nxdomain.
+	if state.QType() == dns.TypeA {
+		switch count {
+		case 0: // cluster only
+			r.cluster = segs[count]
+		case 1: // cluster and hostname
+			r.cluster = segs[count]
+			r.hostname = segs[count-1]
+			r.hostname = segs[count-1]
```
Comment (tpantelis) [inline plugin/lighthouse/parse.go]: Duplicate line.

```suggestion
```

File: pkg/service/controller.go [hunk 13]
```diff
@@ -84,24 +87,41 @@ func (c *Controller) Stop() {
...
 	}
 
 	if !exists {
-		return "", false
+		return nil, false
 	}
 
 	svc := obj.(*v1.Service)
 
+	var mcsServicePort []mcsv1a1.ServicePort
```
Comment (tpantelis) [inline pkg/service/controller.go]:
```suggestion
	var mcsServicePorts []mcsv1a1.ServicePort
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/service/controller.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/service/controller.go]: Done.

File: pkg/serviceimport/map.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/serviceimport/map.go:31]: Done.

File: pkg/serviceimport/map.go (hunk 3 above)
Comment (aswinsuryan) [reply pkg/serviceimport/map.go]: Done.

File: plugin/lighthouse/handler.go (hunk 4 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler.go]: Done.

File: plugin/lighthouse/handler.go (hunk 5 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler.go]: Done.

File: plugin/lighthouse/handler.go (hunk 9 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler.go]: Done.

File: plugin/lighthouse/parse.go (hunk 12 above)
Comment (aswinsuryan) [reply plugin/lighthouse/parse.go]: Done.

File: pkg/service/controller.go (hunk 13 above)
Comment (aswinsuryan) [reply pkg/service/controller.go]: Done.

File: pkg/serviceimport/map.go (hunk 11 above)
Comment (aswinsuryan) [reply pkg/serviceimport/map.go]: Done.

File: plugin/lighthouse/handler.go (hunk 10 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler.go]: Done.

**Comment by vthapar** [comment]
Can we split SRV handler into a separate file? Would be even better to use a plugin/interface type design similar to way etc/kubernetes plugins are written, but that can be done later.

File: plugin/lighthouse/handler_test.go [hunk 14]
```diff
@@ -88,16 +88,16 @@ func (m *MockClusterStatus) LocalClusterID() string {
 }
 
 type MockLocalServices struct {
-	LocalServicesMap map[string]string
+	LocalServicesMap map[string]*serviceimport.DNSRecord
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go:103]: Also add UTs for SRV queries.

**Review by vthapar — CHANGES_REQUESTED** [review]

**Review by mangelajo — DISMISSED** [review]
test

**Comment by mangelajo** [comment]
> test

sorry, please ignore :-/

**Comment by aswinsuryan** [comment]
> Can we split SRV handler into a separate file? Would be even better to use a plugin/interface type design similar to way etc/kubernetes plugins are written, but that can be done later.

Moved record creation to a different file.

File: plugin/lighthouse/handler_test.go (hunk 14 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler_test.go:103]: Added UTs

File: plugin/lighthouse/handler_test.go [hunk 15]
```diff
@@ -172,46 +204,87 @@ func testWithoutFallback() {
...
 		})
+		It("of Type SRV should succeed and write an SRV record response", func() {
+			lh.serviceImports.Put(newServiceImport(namespace2, service1, clusterID, serviceIP, portName1, portNumber1,
+				protcol1, mcsv1a1.ClusterSetIP))
+			executeTestCase(lh, rec, test.Case{
+				Qname: service1 + "." + namespace2 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace2 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]:
```suggestion
					test.SRV(service1 + "." + namespace2 + ".svc.clusterset.local.    5    IN    SRV 0 50 "+ strconv.itoa(portNumber1) +
```

File: plugin/lighthouse/handler_test.go [hunk 16]
```diff
@@ -362,12 +477,24 @@ func testClusterStatus() {
 				},
 			})
 		})
+		It("should succeed and write that cluster's IP as SRV record response", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]:
```suggestion
					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 " + strconv.itoa(portNumber2)
```

File: plugin/lighthouse/handler_test.go [hunk 17]
```diff
@@ -379,6 +506,17 @@ func testClusterStatus() {
 				},
 			})
 		})
+		It("should succeed and write that cluster's IP as SRV record response", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]: Same as above. Use `strconv.itoa(portNumber1)`

File: plugin/lighthouse/handler_test.go [hunk 18]
```diff
@@ -608,6 +799,17 @@ func testLocalService() {
 				},
 			})
 		})
+		It("should succeed and write remote cluster's IP as SRV record response", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(clusterID2 + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]: Same as above

File: plugin/lighthouse/handler_test.go [hunk 19]
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
 				},
 			})
 		})
+		It("should succeed and write remote cluster's IP as SRV record response", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]: same as above

File: plugin/lighthouse/handler_test.go [hunk 20]
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
...
+	})
+
+	When("DNS query of type SRV", func() {
+		It("without portName should return all the ports", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]: Same as above

File: plugin/lighthouse/handler_test.go [hunk 21]
```diff
@@ -627,6 +829,112 @@ func testLocalService() {
...
+		It("with  cluster name should return all the ports from the cluster", func() {
+			executeTestCase(lh, rec, test.Case{
+				Qname: clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local.",
+				Qtype: dns.TypeSRV,
+				Rcode: dns.RcodeSuccess,
+				Answer: []dns.RR{
+					test.SRV(clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 53 " +
+						clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local."),
+					test.SRV(clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local.    5    IN    SRV 0 50 8080 " +
+						clusterID + "." + service1 + "." + namespace1 + ".svc.clusterset.local."),
```
Comment (vthapar) [inline plugin/lighthouse/handler_test.go]: Same as above

**Review by vthapar — CHANGES_REQUESTED** [review]

**Review by vthapar — APPROVED** [review]

**Review by tpantelis — APPROVED** [review]

**Comment by aswinsuryan** [comment]
test

**Pull request merged**